File: Algorithm/284. Peeking Iterator.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-09-09 14:12:14
# @Last modified by:   WuLC
# @Last Modified time: 2016-09-09 14:21:33
# @Email: liangchaowu5@gmail.com

# Below is the interface for Iterator, which is already defined for you.
#
# class Iterator(object):
#     def __init__(self, nums):
#         """
#         Initializes an iterator object to the beginning of a list.
#         :type nums: List[int]
#         """
#
#     def hasNext(self):
#         """
#         Returns true if the iteration has more elements.
#         :rtype: bool
#         """
#
#     def next(self):
#         """
#         Returns the next element in the iteration.
#         :rtype: int
#         """

class PeekingIterator(object):
    def __init__(self, iterator):
        """
        Initialize your data structure here.
        :type iterator: Iterator
        """
        self.iterator = iterator
        self.peak = self.iterator.next() if self.iterator.hasNext() else  None
        

    def peek(self):
        """
        Returns the next element in the iteration without advancing the iterator.
        :rtype: int
        """
        return self.peak
        

    def next(self):
        """
        :rtype: int
        """
        tmp = self.peak
        self.peak = self.iterator.next() if self.iterator.hasNext() else  None
        return tmp

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.peak != None
        

# Your PeekingIterator object will be instantiated and called as such:
# iter = PeekingIterator(Iterator(nums))
# while iter.hasNext():
#     val = iter.peek()   # Get the next element but not advance the iterator.
#     iter.next()         # Should return the same value as [val].

File: Algorithm/Java/284. Peeking Iterator.java
/**
* Author: WuLC
* Date:   2017-04-28 11:37:12
* Last modified by:   WuLC
* Last Modified time: 2017-04-28 11:38:06
* Email: liangchaowu5@gmail.com
*/


// Java Iterator interface reference:
// https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html


// use ArrayList as a queue
class PeekingIterator implements Iterator<Integer> 
{
    private List<Integer> queue;
    public PeekingIterator(Iterator<Integer> iterator) 
    {
        queue = new ArrayList<Integer>();
        while(iterator.hasNext())
        {
            queue.add(iterator.next());
        }
    }

    // Returns the next element in the iteration without advancing the iterator.
    public Integer peek() 
    {
        if (queue.size() > 0) return queue.get(0);
        else return null;
    }

    // hasNext() and next() should behave the same as in the Iterator interface.
    // Override them if needed.
    @Override
    public Integer next() 
    {
        Integer head = queue.get(0);
        queue.remove(0);
        return head;
    }

    @Override
    public boolean hasNext() 
    {
        return queue.size() > 0;
        
    }
}